compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -262,14 +274,80 @@ void setMMALayout(Value aMatrix, Value bMatrix, Value cMatrix,
   setLayout(cMatrix, MMAMatrixType::CMatrix, "cMatrix");
 }
 
+void propagateLayoutToReductions(vector::MultiDimReductionOp reductionOp,
+                                 DenseMap<Value, Layout> &layoutMap) {
+  Value src = reductionOp.getSource();
+  if (!layoutMap.count(src)) return;
+  // Determine layout of result (after reduction)
+  Value result = reductionOp.getResult();
+  Layout resultLayout(layoutMap.at(src));
+  auto reductionDims = reductionOp.getReductionDims().getAsRange<IntegerAttr>();
+  for (auto ia : reductionDims) {
+    resultLayout.reduceDim(ia.getInt());
+  }
+  layoutMap.try_emplace(result, resultLayout);
+  // Propagate result layout to accumulator
+  Value acc = reductionOp.getAcc();
+  layoutMap.try_emplace(acc, resultLayout);
+  layoutMap.at(result).debugPrint("reduced");
+}
+
+void propagateLayoutToBroadcasts(vector::BroadcastOp broadcastOp,
+                                 DenseMap<Value, Layout> &layoutMap) {
+  Value src = broadcastOp.getSource();
+  if (!layoutMap.count(src)) return;
+  // For now, only compute for broadcasts after reductions
+  auto reductionOp = src.getDefiningOp<vector::MultiDimReductionOp>();
+  if (!reductionOp) return;
+  Value reductionSrc = reductionOp.getSource();
+  if (!layoutMap.count(reductionSrc)) return;
+  // For now, require that result shape and source shape are the same
+  // Unless there is a transpose that changes the result shape to the source
+  // shape
+  Value result = broadcastOp.getResult();
+  ArrayRef<int64_t> resultShape =
+      result.getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> srcShape =
+      reductionSrc.getType().cast<ShapedType>().getShape();
+  bool success = srcShape == resultShape;
+  if (!success) {
+    for (Operation *user : result.getUsers()) {
+      if (auto transposeOp = dyn_cast<vector::TransposeOp>(user)) {
+        Value transposedResult = transposeOp.getResult();
+        ArrayRef<int64_t> transposedShape =
+            transposedResult.getType().cast<ShapedType>().getShape();
+        if (transposedShape == srcShape) {
+          // We will lower reduction + broadcast + transpose at once
+          // So assign all their results the same layout
+          layoutMap.try_emplace(transposedResult, layoutMap.at(reductionSrc));
+          layoutMap.at(transposedResult).debugPrint("transposed");
+          break;
+        }
+      }
+    }
+  }
+  if (success) {
+    // Sets layout for either reduction + broadcast (if right shape)
+    // or reduction + broadcast + transpose
+    layoutMap.try_emplace(result, layoutMap.at(reductionSrc));
+    layoutMap.at(result).debugPrint("broadcasted");
+  }
+}
```

Comment:
what about this case don't you want to handle reduction+inner_broadcast? In which case the layout of the result can be the exact same as the reduction source.

Reply:
Right now if reduction + broadcast produces the same shape as the reductionSrc, we set the layout. This would be reduction + outer_broadcast. When we do reduction + inner_broadcast, the shapes of the reductionSrc and the broadcast result are transposed so their layouts wouldn't be the same right? So if the reductionSrc had col = [vecIdy, Laneidx], row = [vecidx, laneidy], then the broadcast would have them flipped?

Reply:
I guess it depends if the reduction is happening on the inner dim or the outter dim. In the IR example you have reductions are on the inner dim so the layout would be the same for a reduction+inner_broadcast

Reply:
Yes, this code is handling both the outer reduce + outer broadcast and inner reduce + inner broadcast conditions.

Reply:
I have updated this function as per our discussion today.